### PR TITLE
Added shift clicks and more!

### PR DESCRIPTION
**Pull request by kaduvert**
+ mode 1 (shift clicks)
+ mode 2 (number clicks, or what its called)
+ mode 4 (drop click aka q press on slot)
+ ts definitions (may contain errors because i have never done this before)

tested changes thoroughly but i don't know if its entirely bugfree.
If you find anything, please let me know and i will fix it!

also this does NOT prevent you from for example put a dirt in the bottom of a brewing stand where you can normally only put a bottle, this was to much work for me and i hope it isn't necessary as it also was not included in previous versions of this project.

### CODE REVIEW

**Comment by u9g** [comment]
do you plan to add tests? i know there aren't yet but could probably add something with the slotupdate(?) event

**Comment by rom1504** [comment]
That's pretty cool.
Could you try to use this with mineflayer (you can use a version like "KadaverBrutalo10/prismarine-windows" to use this in mineflayer ) ? 
I think that could be useful to find possible issues.

Ideally having tests here would also be quite useful.

**Comment by kaduvert** [comment]
i have no real idea of how to implement tests, i tested it on a local minecraft server and checked if the slots matched the real ones.
also i only tested these changes on a 1.8.9 server i do not know if it works on other versions where maybe shift clicking was changed, sorry.
If somebody else could make tests that would be great because as i said i do not know how to

made a [mineflayer fork](https://github.com/KadaverBrutalo10/mineflayer) to use with these changes

**Comment by rom1504** [comment]
Can you PR the mineflayer fork?

**Comment by kaduvert** [comment]
did it

**Comment by kaduvert** [comment]
So, what now?
i've looked a bit deeper and now know how to test it without mineflayer, but what do you mean by test? just some file one can run or this github action thing which tests the code every commit or so?
I would be willing to implement tests but don't know what exactly is meant, someone please explain.

**Comment by rom1504** [comment]
can you run `npm run fix` to fix the lint ?

File: lib/Window.js [hunk 1]
```diff
@@ -118,7 +118,7 @@ module.exports = (Item) => {
       const itemAtHotbarSlot = this.slots[hotbarSlot]
       if (item) {
         if (itemAtHotbarSlot) {
-          if (this.type === 'minecraft:inventory'/* || mcData.isNewerOrEqualTo('1.9')*/) {
+          if (this.type === 'minecraft:inventory'/* || mcData.isNewerOrEqualTo('1.9') */) {
```
Comment (kaduvert) [inline lib/Window.js]: someone with more knowledge about PrismarineJS please implement this as it's needed to provide full support for number clicks in 1.9 and above.

**Comment by kaduvert** [comment]
this updates ts definitions to be up-to-date with https://github.com/PrismarineJS/prismarine-windows/pull/30

File: index.js [hunk 2]
```diff
@@ -67,8 +67,7 @@ function loader (mcVersion) {
 
   return {
     createWindow: (id, type, title, slotCount = undefined) => {
-      let winData = windowByType.get(type)
-      if (!winData) winData = windows[type]
+      let winData = windowByType.get(type) || windows[type]
```
Comment (constantins2001) [inline index.js:70]: What purpose has the added ``|| windows[type]``? The map ``windowByType`` should already have the contents of ``windows``

File: lib/Window.js [hunk 3]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
     }
 
     acceptClick (click) {
-      assert.ok(click.mouseButton === 0 || click.mouseButton === 1)
+      assert.ok(((click.slot >= 0 && click.slot < this.inventoryEnd) || click.slot === -999) && click.mouseButton >= 0 && click.mouseButton <= 6, 'invalid operation')
+
+      switch (click.mode) {
+        case 0:
+          assert.ok(click.mouseButton <= 2, 'invalid operation')
```
Comment (constantins2001) [inline lib/Window.js]: Would replace with ``assert.ok(click.mouseButton === 0 || click.mouseButton === 1, 'invalid button for mode')``

File: lib/Window.js [hunk 4]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
...
+          this.mouseClick(click)
+          break
+
+        case 1:
+          assert.ok(click.mouseButton <= 2, 'invalid operation')
```
Comment (constantins2001) [inline lib/Window.js]: Would replace with ``assert.ok(click.mouseButton === 0 || click.mouseButton === 1, 'invalid button for mode')``

File: lib/Window.js [hunk 5]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
...
+          this.shiftClick(click)
+          break
+
+        case 2:
+          assert.ok(click.mouseButton <= 8 && !this.selectedItem, 'invalid operation')
```
Comment (constantins2001) [inline lib/Window.js]: Would replace with
```
assert.ok(click.mouseButton <= 8 || click.mouseButton >= 0, 'invalid button for mode')
assert.ok(!this.selectedItem, 'invalid operation')
```

File: lib/Window.js [hunk 6]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
...
+          this.numberClick(click)
+          break
+
+        case 4:
+          assert.ok(click.mouseButton < 2 && !this.selectedItem, 'invalid operation')
```
Comment (constantins2001) [inline lib/Window.js]: Would replace with
```
assert.ok(click.mouseButton === 0 || click.mouseButton === 1, 'invalid button for mode')
assert.ok(!this.selectedItem, 'invalid operation')
```

File: lib/Window.js [hunk 7]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
     }
 
     acceptClick (click) {
```
Comment (constantins2001) [inline lib/Window.js:25]: To lazy to PR into your repo, please add the following JSDoc
```
    /**
     * @typedef Click
     * @type {object}
     * @property {number} slot
     * @property {number} mouseButton
     * @property {import("prismarine-item").Item} [item]
     */

    /**
     * Validated and propagates clicks
     * @param click {Click}
     */
```

File: lib/Window.js [hunk 8]
```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
...
+    }
+
+    mouseClick (click) {
       if (click.slot === -999) {
-        this.acceptOutsideWindowClick(click)
-      } else if (click.slot >= this.inventoryStart && click.slot < this.inventoryEnd) {
-        this.acceptInventoryClick(click)
-      } else if (click.slot === this.craftingResultSlot) {
-        this.acceptCraftingClick(click)
+        this.dropSelectedItem(!click.mouseButton)
```
Comment (constantins2001) [inline lib/Window.js:57]: I'd recommend replacing ``!click.mouseButton`` with ``click.mouseButton === 0`` (also line 59) as it's much more understandable :)

**Review by constantins2001 — COMMENTED** [review]
Some ideas that came to mind while reading your code, other than some minor details everything seems good tho ;)
(And should you have some free time to improve code readability, I'd suggest adding some [JSDoc](https://devhints.io/jsdoc) )

PS: Will add some more comments in the near future, running out of time for today :/

File: lib/Window.js (hunk 1 above)
Comment (extremeheat) [reply lib/Window.js]: Can you explain what number clicks are in this context? What exactly are you trying to do here?

**Review by extremeheat — COMMENTED** [review]
What's the status of the PR? Are you able to finish work on this? Would be helpful for GUIs. The only important thing I think you need to fix is the 1.9 mcdata issue:

**Comment by kaduvert** [comment]
Short answer: it's maybe 75% done. i currently don't have time to finish and am frustrated with the cooperation between maintainers and me.
number clicks means moving your cursor on a slot and pressing a number, i did not have a better name, if you have: please tell me.
what i am trying to do there is: if you have a chest open and execute this type of click it handles it differently in 1.9 and above as in 1.8. more specific: if slot where your cursor is contains item and the slot where it should go in 1.8 searches for new place for the item in the hotbar slot as in 1.9 it just swaps items also in 1.8 this only happens in containers which are not your inventory.

Long answer:
i have finished some work over the past few weeks (also the mcData one) but i currently don't have time to finish completely.
upon opening this pr i thought that in some inventorys some slots only take certain items wasn't a big thing, because you had to do it with intent. but as i continued i noticed that if you shift clicked for example armor around in your normal inventory it goes into the armor slots which leads to desync. i at the time don't really have time for this as it's a bigger thing to do (specify which slots in which container take what items)
I will likely do this but for now i'm not working on this as i've also lost motivation, mainly because the cooperation between maintainers and me is poor. To me it seems as no maintainer of this project is really interested in this, and the cooperation with them is as i said also not what i would want to work with.

**Comment by extremeheat** [comment]
Yes, you're right, the click-based inventory protocol is pretty poor and heavily linked to the Minecraft codebase. As for the issues with shift clicking for armor slots, I believe it should be simple enough to add some special rules for armor items, or even to reject these transactions to not desynchronize the transactions. This is important to support for minecraft-inventory-gui/prismarine-viewer, so I'll look at this later if you're not able to.

**Comment by kaduvert** [comment]
It's not like i'm too dumb too implement and i have a rough idea of how to do it but don't find the time and am not motivated, maybe we can talk sometime over discord or something and do it together if you want to.
You don't only have do create special rules for inventory but for all containers.
Rejecting transactions is not an option for me, sorry.

**Comment by rom1504** [comment]
I don't feel comfortable merging this complex change without any test.
@extremeheat do you think this is likely to work ? do you think we should merge ?

**Comment by kaduvert** [comment]
from sometime ago i had made tests for the clicks, these can be run by just ```node test.js```, as you will see the armor shift click fails, that is what still has to be implemented such as the mcVersion thing mentioned earlier this conversation

@rom1504 i would not merge yet, if you or someone of your team could help me implement the things above, that would speed up development. i am at i point where i can't move forward confidentially without someone with more knowledge about PrismarineJS in general who can help me out on a few things.

**Comment by extremeheat** [comment]
Yeah, I agree with the need for tests. I see you added a file, you can see how the other projects use mocha or jest and do the same here. Make the test run on the npm test script that way we can see the results of the test on CI. If you need help you can join the #contribute room on discord to discuss. But I don’t get why you can’t reject the unimplemented transactions, that would avoid de sync issues. As for the shit clicking issue, I’d make a bot try and iterate through all the items and shift click them all and log how the server responds to them and verify what slot gets updated. Then it can be implemented here. That’s assuming it’s not just armor or other edge cases that have shift clicking problems.

**Comment by kaduvert** [comment]
ik what you mean, but i like to completely implement shift-clicks straight away rather than having this annoying workaround which eventually has to be exchanged for a full implementation.
thanks for the hint with the discord channel, i will stop by soon when i'm willing to work on this. ;)

**Comment by rom1504** [comment]
closed due to no tests, tracked there https://github.com/PrismarineJS/prismarine-windows/issues/43 if someone want to continue in the future

File: test.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
+const windows = require('./')('1.8')
```
Comment (rom1504) [inline test.js:1]: I see there is a test.js now...

would just need to convert it to mocha then

Comment (kaduvert) [reply test.js:1]: made tests with jest because mocha didn't work for me, hope thats fine.

Comment (extremeheat) [reply test.js:1]: You forgot to commit and push

Comment (kaduvert) [reply test.js:1]: no, i did not.
now that the pr is closed i guess that commits to my forked repo don't show up here anymore, but they can be seen here:
https://github.com/KadaverBrutalo10/prismarine-windows

**Comment by 182exe** [comment]
🙏

**Comment by kaduvert** [comment]
> 🙏 

do not be fooled by the closed status, i've got it implemented. see #74

**Comment by 182exe** [comment]
you damn legend

**Pull request closed without merging**